### PR TITLE
time.Time support

### PR DESCRIPTION
**Pull request by Shixzie**
time.Time support, changed error messages for better consistency as well as some nil checks that were missing

### CODE REVIEW

**Comment by guduxingzou** [comment]
@Shixzie ths for your work,it's cool!

**Comment by Shixzie** [comment]
@kniren please review it :); and @guduxingzou , thanks to you :+1:

File: series/series.go [hunk 1]
```diff
@@ -530,20 +582,26 @@ func (s Series) Len() int {
...
+	}
 	var ret []string
 	// If name exists print name
 	if s.Name != "" {
 		ret = append(ret, "Name: "+s.Name)
 	}
-	ret = append(ret, "Type: "+fmt.Sprint(s.t))
-	ret = append(ret, "Length: "+fmt.Sprint(s.Len()))
+	ret = append(ret, fmt.Sprintf("Type: %v", s.t))
+	ret = append(ret, fmt.Sprintf("Length: %v", s.Len()))
```
Comment (kniren) [inline series/series.go]: Change the `%v` to `%d` for consistency with the rest of the library

File: series/series.go [hunk 2]
```diff
@@ -530,20 +582,26 @@ func (s Series) Len() int {
...
+		return s.Err.Error()
+	}
 	var ret []string
 	// If name exists print name
 	if s.Name != "" {
 		ret = append(ret, "Name: "+s.Name)
 	}
-	ret = append(ret, "Type: "+fmt.Sprint(s.t))
-	ret = append(ret, "Length: "+fmt.Sprint(s.Len()))
+	ret = append(ret, fmt.Sprintf("Type: %v", s.t))
```
Comment (kniren) [inline series/series.go]: Change the `%v` to `%s` for consistency with the rest of the library

File: series/series.go [hunk 3]
```diff
@@ -583,7 +641,7 @@ func parseIndexes(l int, indexes Indexes) ([]int, error) {
 	case []bool:
 		bools := indexes.([]bool)
 		if len(bools) != l {
-			return nil, fmt.Errorf("indexing error: index dimensions mismatch")
+			return nil, fmt.Errorf("series.parseIndexes(): index dimensions mismatch")
```
Comment (kniren) [inline series/series.go]: It was good before, revert to it

File: series/series_test.go [hunk 4]
```diff
@@ -94,6 +95,18 @@ func TestSeries_Compare(t *testing.T) {
 			Bools([]bool{true, false, true}),
 		},
 		{
+			Times([]time.Time{time.Now(), time.Now().Add(time.Minute * 2)}),
```
Comment (kniren) [inline series/series_test.go]: This is extremely dangerous! `time.Now()` will change on every call! Parse a fixed date from a `string` instead so that every time tested is the same

File: series/series_test.go [hunk 5]
```diff
@@ -381,6 +454,18 @@ func TestSeries_Compare(t *testing.T) {
...
+		{
+			Times([]time.Time{time.Now(), time.Now().Add(time.Hour * 240)}),
+			In,
+			"01/25/2017",
+			Bools([]bool{false, false}),
+		},
+		{
+			Times([]time.Time{time.Now(), time.Now().Add(time.Hour * -240)}),
+			In,
+			[]string{"25/01/2017", "01/31/2017"},
```
Comment (kniren) [inline series/series_test.go]: This is not the right time format to be parsed!

From the docs:

`RFC3339     = "2006-01-02T15:04:05Z07:00"`

File: series/series_test.go [hunk 6]
```diff
@@ -449,6 +534,36 @@ func TestSeries_Subset(t *testing.T) {
 			Bools([]bool{true, false, false, true, true}),
 			"[A K D]",
 		},
+		{
```
Comment (kniren) [inline series/series_test.go]: If the PR is about adding support to `time.Time` try not to add things that have nothing to do with it. These tests are nice to have though, but please submit a separate PR next time.

File: series/series_test.go [hunk 7]
```diff
@@ -1165,3 +1285,49 @@ func TestSeries_IsNaN(t *testing.T) {
...
+func TestSeries_Empty(t *testing.T) {
+	tests := []struct {
+		name string
+		s    Series
+		want Series
+	}{
+		{"Series Empty", Series{Name: "test", t: String}, New([]int{}, String, "test")},
+	}
+	for i, tt := range tests {
+		t.Run(tt.name, func(t *testing.T) {
```
Comment (kniren) [inline series/series_test.go]: Please, try to be consistent with the way the rest of the tests are written. Don't use `t.Run`. Also, this should be a separate PR. (Same thing with TestSeries_Append)

File: series/type-bool.go [hunk 8]
```diff
@@ -19,15 +20,12 @@ func (e *boolElement) Set(value interface{}) {
 			e.nan = true
 			return
 		}
-		switch strings.ToLower(value.(string)) {
-		case "true", "t", "1":
-			e.e = true
-		case "false", "f", "0":
-			e.e = false
-		default:
+		v, err := strconv.ParseBool(value.(string))
```
Comment (kniren) [inline series/type-bool.go]: I agree with this change and will merge it here, but again, this should be a separate PR in the future

File: series/type-time.go [hunk 9]
```diff
@@ -0,0 +1,150 @@
+package series
+
+import (
+	"fmt"
+	"time"
+)
+
+// timeElement is the concrete implementation of the Element interface for
+// time.Time. If the stored time.Time is zero, it will be considered as a NaN
+// element.
```
Comment (kniren) [inline series/type-time.go]: Since the zero time will be considered NaN, there is no need to have a separate `nan` field on the struct. Furthermore, after removing it the rest of the code should be adjusted to match the change.

Maybe we could do `type timeElement time.Time` instead, since we only need one field

File: series/type-time.go [hunk 10]
```diff
@@ -0,0 +1,150 @@
...
+		var err error
+		e.e, err = time.Parse(timeformat, v)
+		if err != nil {
+			e.e = time.Time{}
+		}
+		return
+	case int, float64, bool:
+		e.e = time.Time{}
+	case time.Time:
+		e.e = value.(time.Time)
```
Comment (kniren) [inline series/type-time.go]: Are we sure that we are making a copy of `time.Time` instead of passing a reference? Need to check if it's necessary to instantiate a new `time.Time` instead...

File: series/type-time.go [hunk 11]
```diff
@@ -0,0 +1,150 @@
...
+		return "NaN"
+	}
+	return e.e.Format(timeformat)
+}
+
+func (e timeElement) Int() (int, error) {
+	if e.IsNA() {
+		return 0, fmt.Errorf("timeElement.Int(): can't convert NaN to int")
+	}
+	return int(e.e.Unix()), nil
```
Comment (kniren) [inline series/type-time.go]: What happens here if e.e.Unix overflows int? (int is 32b instead of 64b) Need to check this before making the conversion and throw an error if necessary.

Also, are we sure we want `Unix` (Seconds from Epoch) or do we want `UnixNano` (Nanoseconds from Epoch)? Maybe we could discuss this on the data-science slack channel...

File: series/type-time.go [hunk 12]
```diff
@@ -0,0 +1,150 @@
...
+}
+
+func (e timeElement) Float() float64 {
+	return float64(e.e.Unix())
```
Comment (kniren) [inline series/type-time.go]: Same as above, maybe we could use e.e.Unix() for Int and `e.e.UnixNano()` for Float? Again, maybe it's best to check on the slack channel

File: series/type-time.go [hunk 13]
```diff
@@ -0,0 +1,150 @@
...
+}
+
+func (e timeElement) Time() (time.Time, error) {
+	if e.IsNA() {
+		return time.Time{}, fmt.Errorf("timeElement.Time(): can't convert NaN to time.Time")
+	}
+	return e.e, nil
+}
+
+func (e timeElement) Addr() string {
```
Comment (kniren) [inline series/type-time.go]: I think we should remove Addr altogether, but don't do it here, for now the implementation requires it, so we will make the change on all elements later

**Comment by Shixzie** [comment]
All requested changes (changes on a separate branch, just time.Time stuff) has been solved on [PR #27](https://github.com/kniren/gota/pull/27)
